### PR TITLE
expose nonparticipation txn field in teal

### PR DESCRIPTION
**Pull request by shiqizng**
## Summary
Expose the field Nonparticipation from KeyregTxnFields in teal. This is needed for checking keyreg txns' nonparticipation value in logicsigns. 

## Test Plan
Updated tests in logic

### CODE REVIEW

File: data/transactions/logic/doc.go [hunk 1]
```diff
@@ -337,6 +337,7 @@ var txnFieldDocs = map[string]string{
 	"VoteFirst":                "The first round that the participation key is valid.",
 	"VoteLast":                 "The last round that the participation key is valid.",
 	"VoteKeyDilution":          "Dilution for the 2-level participation key",
+	"Nonparticipation":         "Flag for participation key",
```
Comment (jasonpaulos) [inline data/transactions/logic/doc.go:340]: A better description might be "Key registration participation flag"

File: data/transactions/logic/fields.go [hunk 2]
```diff
@@ -232,6 +235,7 @@ var txnFieldSpecs = []txnFieldSpec{
 	{LocalNumUint, StackUint64, 3},
 	{LocalNumByteSlice, StackUint64, 3},
 	{ExtraProgramPages, StackUint64, 4},
+	{Nonparticipation, StackUint64, 0},
```
Comment (jasonpaulos) [inline data/transactions/logic/fields.go:238]: Unfortunately I think we'll need to make this first available in v5

**Review by jasonpaulos — COMMENTED** [review]
For the most part looks good, except the new field should be enabled only for v5 and onward.

Also, we may want to make it such that LogicSigs running programs v4 and lower automatically reject transactions that contain true for the nonparticipation field, since they cannot inspect it. A similar thing was done when we introducing rekeying in TEAL v2. I'll see if I can find any relevant code.

edit: the relevant code is `ComputeMinTealVersion` in `data/transactions/logic/eval.go`. But since it's currently valid for programs to execute in the same group as nonparticipation keyreg txns, we'd probably need to introduce a new consensus param to make this change. I'm not sure if that would be worth it.

**Comment by shiqizng** [comment]
> For the most part looks good, except the new field should be enabled only for v5 and onward.
> 
> Also, we may want to make it such that LogicSigs running programs v4 and lower automatically reject transactions that contain true for the nonparticipation field, since they cannot inspect it. A similar thing was done when we introducing rekeying in TEAL v2. I'll see if I can find any relevant code.
> 
> edit: the relevant code is `ComputeMinTealVersion` in `data/transactions/logic/eval.go`. But since it's currently valid for programs to execute in the same group as nonparticipation keyreg txns, we'd probably need to introduce a new consensus param to make this change. I'm not sure if that would be worth it.

I don't think it would be worth a consensus upgrade either.

File: data/transactions/logic/eval.go [hunk 3]
```diff
@@ -1886,6 +1886,12 @@ func (cx *evalContext) txnFieldToStack(txn *transactions.Transaction, field TxnF
 		sv.Uint = uint64(txn.VoteLast)
 	case VoteKeyDilution:
 		sv.Uint = txn.VoteKeyDilution
+	case Nonparticipation:
+		if txn.Nonparticipation {
+			sv.Uint = 1
+		} else {
+			sv.Uint = 0
+		}
```
Comment (jannotti) [inline data/transactions/logic/eval.go:1894]:
```suggestion
		sv.Uint = boolToUint(txn.Nonparticipation)
```

File: data/transactions/logic/doc.go [hunk 4]
```diff
@@ -337,6 +337,7 @@ var txnFieldDocs = map[string]string{
 	"VoteFirst":                "The first round that the participation key is valid.",
 	"VoteLast":                 "The last round that the participation key is valid.",
 	"VoteKeyDilution":          "Dilution for the 2-level participation key",
+	"Nonparticipation":         "Key registration participation flag",
```
Comment (jannotti) [inline data/transactions/logic/doc.go:340]: How about "Marks an account nonparticipating for rewards"?

That is similar to the language we use here: https://developer.algorand.org/docs/reference/transactions/

**Review by jannotti — CHANGES_REQUESTED** [review]
Looks good. Two small recomendations.

**Review by jasonpaulos — APPROVED** [review]

**Review by jannotti — APPROVED** [review]

**Pull request merged**